internal/tiltfile2/tiltfile_state.go

```diff
@@ -0,0 +1,292 @@
+		if !assembledImages[k] {
+			return nil, fmt.Errorf("image %v is not used in any resource", k)
+		}
+	}
+
+	return s.k8s, nil
+}
+
+func (s *tiltfileState) validateK8s(r *k8sResource, assembledImages map[string]bool) error {
+	var images []reference.Named
+	for _, e := range r.k8s {
+		entityImages, err := e.FindImages()
+		if err != nil {
+			return err
+		}
+		images = append(images, entityImages...)
+	}
+
+	for _, image := range images {
+		if _, ok := s.imagesByName[image.Name()]; !ok {
+			continue
+		}
+		if r.imageRef == "" {
+			r.imageRef = image.Name()
+		} else if r.imageRef == image.Name() {
+			continue
+		} else {
+			return fmt.Errorf("resource %q contains two images to be built: %q, %q. You can use k8s_yaml to include a lot of yaml and then Tilt will create resources automatically. If this doesn't solve your case (e.g. you have one pod that has two images that Tilt updates), please reach out so we can understand and prioritize", r.name, r.imageRef, image.Name())
+		}
+	}
+
+	if len(r.k8s) == 0 && r.name != unresourcedName {
+		if r.imageRef == "" {
+			return fmt.Errorf("resource %q: no matching resource", r.name)
+		}
+		return fmt.Errorf("resource %q: could not find image %q; perhaps there's a typo?", r.name, r.imageRef)
+	}
+
+	assembledImages[r.imageRef] = true
+
+	return nil
+}
+
+func (s *tiltfileState) findExpandTarget(image reference.Named) (*k8sResource, error) {
+	// first, match an empty resource that has this exact imageRef
+	for _, r := range s.k8s {
+		if len(r.k8s) == 0 && r.imageRef == image.Name() {
+			return r, nil
+		}
+	}
+
+	// next, match an empty resource that has the same name
+	name := filepath.Base(image.Name())
+	for _, r := range s.k8s {
+		if len(r.k8s) == 0 && r.name == name {
+			return r, nil
+		}
+	}
+
+	// otherwise, create a new resource
+	return s.makeK8sResource(name)
+}
+
+func (s *tiltfileState) findUnresourcedImages() ([]reference.Named, error) {
+	var result []reference.Named
+	seen := make(map[string]bool)
+
+	for _, e := range s.k8sUnresourced {
+		images, err := e.FindImages()
+		if err != nil {
+			return nil, err
+		}
+		var entityImages []reference.Named
+		for _, image := range images {
+			if _, ok := s.imagesByName[image.Name()]; ok {
+				entityImages = append(entityImages, image)
+			}
+		}
+		if len(entityImages) == 0 {
+			continue
+		}
+		if len(entityImages) > 1 {
+			str, err := k8s.SerializeYAML([]k8s.K8sEntity{e})
+			if err != nil {
+				str = err.Error()
+			}
+			return nil, fmt.Errorf("Found an entity with multiple images registered with k8s_yaml. Tilt doesn't support this yet; please reach out so we can understand and prioritize this case. found images: %q, entity: %q.", entityImages, str)
+		}
+		img := entityImages[0]
+		if !seen[img.Name()] {
+			result = append(result, img)
+			seen[img.Name()] = true
+		}
+	}
+	return result, nil
+}
+
+func (s *tiltfileState) extractImage(dest *k8sResource, imageRef reference.Named) error {
```

Comment:
might rename since you're not extracting the IMAGES, you're extracting k8s entities matching the image (and those matching those matching the image...) `extractK8sForImage` or something?